backend/upload/models.py

```diff
@@ -1,3 +1,331 @@
-# from django.db import models
+from pathlib import Path
+import re
+from lxml import etree
+import logging
 
-# Create your models here.
+from django.db import models
+from django.contrib.auth.models import User
+from django.utils.text import slugify
+
+from corpus2alpino.converter import Converter
+from corpus2alpino.collectors.filesystem import FilesystemCollector
+from corpus2alpino.targets.memory import MemoryTarget
+from corpus2alpino.writers.lassy import LassyWriter
+
+from treebanks.models import Treebank, Component, BaseXDB
+from services.alpino import alpino, AlpinoError
+from services.basex import basex
+
+logger = logging.getLogger(__name__)
+
+MAXIMUM_DATABASE_SIZE = 1024 * 1024 * 10  # 10 MiB
+
+
+class UploadError(RuntimeError):
+    pass
+
+
+class TreebankUpload(models.Model):
+    '''Class to upload texts of various input formats to GrETEL. The model
+    can keep information about the upload progress during the various stages
+    of the process (i.e. after uploading and inspecting the files, during
+    processing and after processing), allowing user interaction in between,
+    but if the process can be executed in one run (e.g. after calling a
+    Django management command) it does not have to be saved at all. To use,
+    make sure that input_file or input_dir is set and subsequently call
+    prepare() and process().'''
+    ALPINO = 'A'
+    CHAT = 'C'
+    TXT = 'T'
+    FOLIA = 'F'
+    FORMAT_CHOICES = [
```

Comment:
This could be (marginally) nicer with an enum type: https://docs.djangoproject.com/en/4.1/ref/models/fields/#field-choices-enum-types